File: inference_framework-java/src/main/java/com/inferenceframework/model/ModelInterface.java
package com.inferenceframework.model;  // model module

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.inferenceframework.event.OutputEvent;  // from event module

/**
 * Abstract interface that all models must implement for pluggable inference.
 * Adapted for Java: uses List<Map> instead of Polars DataFrame for batch processing.
 */
public interface ModelInterface {

    /**
     * Process batch of messages (as List<Map>) and corresponding features.
     * Return list of OutputEvent instances (fixed format).
     *
     * @param messages List of validated message maps (from schema/JSON)
     * @param features Dict of user_id -> feature_vector
     * @return List of OutputEvent objects
     */
    List<OutputEvent> process(List<Map<String, Object>> messages, Map<String, Map<String, Object>> features);

    /**
     * Generate standard OutputEvent (models should call this).
     * Factory for consistency (auto ID/timestamp, model_id).
     */
    default OutputEvent generateOutputEvent(String userId, double score, Map<String, Object> data, String modelId) {
        if (data == null) {
            data = new HashMap<>();
        }
        if (modelId == null) {
            modelId = this.getClass().getSimpleName().toLowerCase();
        }
        return OutputEvent.create(userId, score, data, modelId);
    }
}

File: inference_framework-java/archive/python_original/inference_framework/model_interface.py
from abc import ABC, abstractmethod
from typing import Dict, Any, List
import polars as pl
from inference_framework.output_event import OutputEvent

class ModelInterface(ABC):
    """Abstract interface that all models must implement for pluggable inference."""
    
    @abstractmethod
    def process(self, df: pl.DataFrame, features: Dict[str, Dict[str, Any]]) -> List[OutputEvent]:
        """
        Process the Polars DataFrame of messages and corresponding features.
        Return list of OutputEvent instances (fixed format).
        
        Args:
            df: Polars DataFrame with validated messages (columns from schema)
            features: Dict of user_id -> feature_vector
        
        Returns:
            List of OutputEvent objects
        """
        pass
    
    def generate_output_event(self, user_id: str, score: float, data: Dict[str, Any] = None, model_id: str = None) -> OutputEvent:
        """Generate standard OutputEvent (models should call this or use factory)."""
        if data is None:
            data = {}
        if model_id is None:
            model_id = self.__class__.__name__.lower()
        # Use factory for consistency (auto ID/timestamp)
        return OutputEvent.create(user_id=user_id, score=score, data=data, model_id=model_id)
